File: 206. Reverse Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        // base case
        if (head == null || head.next == null) {
            return head;
        }        
        //find the last node
        ListNode prev = head;
        ListNode node = prev.next;
        while (node.next != null) {
            prev = node;
            node = node.next;
        }
        // make the last node the new head
        ListNode newHead = new ListNode(node.val);
        // delete the last node
        prev.next = null;
        // reverse the rest recursively
        newHead.next = reverseList(head);
        return newHead;
    }
}

File: 206. Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        
        # base case
        if not head or not head.next:
            return head
        
        # brutal force
        prev = head
        node = head.next
        while node.next:
            prev = prev.next 
            node = node.next
        # node is tail, prev is the node before tail
        prev.next = None
        node.next = self.reverseList(head)
        return node